### PR TITLE
Add pre-upgrade helm hook job

### PR DESCRIPTION
**Pull request by stonezdj**
The job will migrate database before upgrade
  Add settings of enableMigrateHelmHook to helm chart
  Update readme.md to add the enableMigrateHelmHook option

### CODE REVIEW

File: templates/core/core-pre-upgrade-job.yaml [hunk 1]
```diff
@@ -0,0 +1,173 @@
+{{- if .Values.core.enableMigrateHelmHook }}
```
Comment (reasonerjt) [inline templates/core/core-pre-upgrade-job.yaml:1]: Many attributes in the job are copied from the core pod, but are not needed, they should be removed?  IMO the job should be much simpler.

Comment (stonezdj) [reply templates/core/core-pre-upgrade-job.yaml:1]: Updated

File: templates/core/core-pre-upgrade-job.yaml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    {{- with .Values.core.affinity }}
+      affinity:
+{{ toYaml . | indent 8 }}
+    {{- end }}
+    {{- with .Values.core.tolerations }}
+      tolerations:
+{{ toYaml . | indent 8 }}
+    {{- end }}
+    {{- if .Values.core.priorityClassName }}
+      priorityClassName: {{ .Values.core.priorityClassName }}
```
Comment (reasonerjt) [inline templates/core/core-pre-upgrade-job.yaml:71]: is this needed?

File: README.md [hunk 3]
```diff
@@ -223,6 +223,7 @@ The following table lists the configurable parameters of the Harbor chart and th
...
+| `core.enableMigrateHelmHook`                                                | Run the migration job via helm hook                                                                                                                                                                                                                                                                                                                                                                                                                                                            | `false`                                                                             |
```
Comment (reasonerjt) [inline README.md:226]: I don't think the option should be put under `core`

File: templates/core/core-pre-upgrade-job.yaml [hunk 4]
```diff
@@ -0,0 +1,73 @@
+{{- if .Values.core.enableMigrateHelmHook }}
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: {{ template "harbor.core" . }}-job
```
Comment (reasonerjt) [inline templates/core/core-pre-upgrade-job.yaml:5]: call it migration job?

File: templates/core/core-pre-upgrade-job.yaml [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  labels:
+{{ include "harbor.labels" . | indent 4 }}
+    component: core
```
Comment (reasonerjt) [inline templates/core/core-pre-upgrade-job.yaml:8]: is it correct?

File: templates/core/core-pre-upgrade-job.yaml [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    "helm.sh/hook": pre-upgrade
+    "helm.sh/hook-weight": "-5"
+spec:
+  template:
+    metadata:
+      labels:
+{{ include "harbor.matchLabels" . | indent 8 }}
+        component: core-job
+      annotations:
+        checksum/configmap: {{ include (print $.Template.BasePath "/core/core-cm.yaml") . | sha256sum }}
```
Comment (reasonerjt) [inline templates/core/core-pre-upgrade-job.yaml:21]: this checksum was introduced to control the recreation of the pod, so probably does not make sense for the job?

File: templates/core/core-pre-upgrade-job.yaml [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        runAsUser: 10000
+        fsGroup: 10000
+{{- if .Values.core.serviceAccountName }}
+      serviceAccountName: {{ .Values.core.serviceAccountName }}
+{{- end -}}
+      {{- with .Values.imagePullSecrets }}
+      imagePullSecrets:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      automountServiceAccountToken: {{ .Values.core.automountServiceAccountToken | default false }}
```
Comment (reasonerjt) [inline templates/core/core-pre-upgrade-job.yaml:35]: This is for calling the api-server?  probably we don't need it in the job? 
@ywk253100 what do you think?

File: README.md (hunk 3 above)
Comment (stonezdj) [reply README.md:226]: updated

File: templates/core/core-pre-upgrade-job.yaml (hunk 4 above)
Comment (stonezdj) [reply templates/core/core-pre-upgrade-job.yaml:5]: Updated

File: templates/core/core-pre-upgrade-job.yaml (hunk 5 above)
Comment (stonezdj) [reply templates/core/core-pre-upgrade-job.yaml:8]: change to migrator

File: templates/core/core-pre-upgrade-job.yaml (hunk 2 above)
Comment (stonezdj) [reply templates/core/core-pre-upgrade-job.yaml:71]: removed

File: templates/core/core-pre-upgrade-job.yaml (hunk 6 above)
Comment (stonezdj) [reply templates/core/core-pre-upgrade-job.yaml]: removed

File: templates/core/core-pre-upgrade-job.yaml (hunk 7 above)
Comment (stonezdj) [reply templates/core/core-pre-upgrade-job.yaml]: remove

**Review by ywk253100 — APPROVED** [review]

**Pull request merged**